### PR TITLE
kinesis implementation

### PR DESCRIPTION
**Pull request by mmniaziqb**

### CODE REVIEW

File: sources/kinesis/kinesiswatcher.go [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+		return up, nil
+	}
+}
+
+var (
+	stream = "/aws/lambda/esp-production"
+)
+
+func New() *Watcher {
+	flag.StringVar(&stream, "kinesis.stream", stream, "The kinesis stream to read from")
```
Comment (tcolgate) [inline sources/kinesis/kinesiswatcher.go]: These become global CLI args. You should pass them in to the New() constructor as args and let them client arrange for themt o be set up properly.

File: sources/kinesis/kinesiswatcher.go [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+	select {
+	case <-ctx.Done():
+		return nil, ctx.Err()
+	case up := <-w.ups:
+		return up, nil
+	}
+}
+
+var (
+	stream = "/aws/lambda/esp-production"
```
Comment (tcolgate) [inline sources/kinesis/kinesiswatcher.go]: This should be an argument to New()

File: sources/kinesis/kinesiswatcher.go [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+)
+
+func New() *Watcher {
+	flag.StringVar(&stream, "kinesis.stream", stream, "The kinesis stream to read from")
+	flag.Set("logtostderr", "true")
```
Comment (tcolgate) [inline sources/kinesis/kinesiswatcher.go]: main of the actual client needs to take care of this and the following Parse()

File: sources/kinesis/kinesiswatcher.go [hunk 4]
```diff
@@ -0,0 +1,99 @@
...
+	flag.Parse()
+
+	return &Watcher{}
+}
+
+func (w *Watcher) initialize() error {
+	awsSession, err := session.NewSession()
```
Comment (tcolgate) [inline sources/kinesis/kinesiswatcher.go]: I'd actually take the AWS session/client as an arg. Probably an interface that has the method set you need.

File: sources/kinesis/kinesissource.go [hunk 5]
```diff
@@ -0,0 +1,84 @@
...
+	})
+
+	if err != nil {
+		return nil, errors.Wrapf(err, "could not get records for shard %v: %v", mr.shardId)
+	}
+
+	mr.shardIterator = resp.NextShardIterator
+
+	for _, r := range resp.Records {
+		message.Text = string(r.Data)
```
Comment (tcolgate) [inline sources/kinesis/kinesissource.go]: You should set a timestamp, preferably one from the original message. You may also want to add some labels, e.g. maybe some kind of shard identified? deifinitely the kenesis stream name.
  If kinesis messages have any other metadata with them, consider adding it as labels.

File: .gitignore [hunk 6]
```diff
@@ -2,3 +2,5 @@ logspray-server
 logspray-reader
 logspray-client
 dist/
+.idea
+logspray.iml
```
Comment (tcolgate) [inline .gitignore]: These ignores could be added you your personal git config file (don't want to add every potential editor temporary/config file in .gitignore.

File: sources/kinesis/kinesiswatcher.go [hunk 7]
```diff
@@ -0,0 +1,94 @@
...
+type Watcher struct {
+	ups     chan []*sources.Update
+	kinesis *kinesis.Kinesis
+	stream  string
+	region  *string
+}
+
+// Next should be called each time you wish to watch for an update.
+func (w *Watcher) Next(ctx context.Context) ([]*sources.Update, error) {
+	if w.ups == nil {
```
Comment (tcolgate) [inline sources/kinesis/kinesiswatcher.go]: Is there no chance that the list of shards / ids will get updated dynamically? That seems a little odd.

File: sources/kinesis/kinesissource.go [hunk 8]
```diff
@@ -0,0 +1,152 @@
...
+			marshlingErr := json.Unmarshal(buf.Bytes(), &kinesisMsg)
+			gzipReader.Close()
+			if marshlingErr != nil {
+				continue
+			}
+
+			for _, log := range kinesisMsg.LogEvents {
+				logsprayMsg := &logspray.Message{}
+				logsprayMsg.Text = parseLog(log)
+				logsprayMsg.Labels = map[string]string{
```
Comment (tcolgate) [inline sources/kinesis/kinesissource.go]: These labels are constant through the whole stream, so set them as target labels (on sources.Update in the watcher)

File: sources/kinesis/kinesissource.go [hunk 9]
```diff
@@ -0,0 +1,152 @@
...
+					"job":             "kinesis",
+					"stream_name":     mr.stream,
+					"stream_shard_id": mr.shardId,
+				}
+
```
Comment (tcolgate) [inline sources/kinesis/kinesissource.go]: You should set a timestamp here, either use time.Now() , or pull a timestamp from the kinesis message (I'd be temtped to use the one on the kinesis message that marks the time it was accepted into kinesis)

File: sources/kinesis/kinesissource.go [hunk 10]
```diff
@@ -0,0 +1,152 @@
...
+				msg := Message{logsprayMsg, nil}
+
+				select {
+				case <-ctx.Done():
+					close(mr.messagesChannel)
```
Comment (tcolgate) [inline sources/kinesis/kinesissource.go]: can this be defer'd at the top of the function?

File: sources/kinesis/kinesissource.go [hunk 11]
```diff
@@ -0,0 +1,152 @@
...
+	return message.logsprayMsg, message.error
+}
+
+func (w *Watcher) shardIterator(ctx context.Context, shardId string, fromStart bool) (*string, error) {
+	var shardIteratorType *string
+	if fromStart {
+		shardIteratorType = aws.String("TRIM_HORIZON")
+	} else {
+		shardIteratorType = aws.String("LATEST")
+	}
```
Comment (tcolgate) [inline sources/kinesis/kinesissource.go]:
```	
        shardIteratorType = aws.String("LATEST")
￼ 	if fromStart {
￼ 		shardIteratorType = aws.String("TRIM_HORIZON")
￼ 	} ```

File: sources/kinesis/kinesiswatcher.go (hunk 7 above)
Comment (mmniaziqb) [inline sources/kinesis/kinesiswatcher.go]: ya shards and id will remain same until unless we dont explicitly do rehsarding

**Comment by tcolgate** [comment]
This looks basically OK. There are a couple of things that could be refactored:
- the limit on message read should probably be an option on the construction of the reader
- the parseMessage thing should probably be able to error (e.g. the typecast could fail), and some more of the logic could possibly be moved in there.
- But otherwise looks OK.

**Pull request merged**